ophyd/areadetector/base.py

```diff
@@ -148,3 +150,117 @@ def default_match(attr, signal, doc):
 
             if match:
                 match_fcn(attr=attr, signal=getattr(self, attr), doc=doc)
+
+    def stage(self, *args, **kwargs):
+        if not self.validate_asyn_ports():
+            missing_plugins = self.missing_plugins()
+            raise RuntimeError('The asyn ports {!r} are used by plugins '
+                               'that ophyd is aware of but the source plugin '
+                               'is not.  Please reconfigure your device to '
+                               'include the source plugin or reconfigure '
+                               'to not use these ports.'
+                               ''.format(missing_plugins))
+        return super().stage(*args, **kwargs)
+
+    def get_plugin_by_asyn_port(self, port_name):
+        '''Get the plugin which has the given asyn port name
+
+        Parameters
+        ----------
+        port_name : str
+            The port name to search for
+
+        Returns
+        -------
+        ret : ADBase or None
+            Either the requested plugin or None if not found
+
+        '''
+        try:
+            name = self.port_name.get()
+            if name == port_name:
```

Comment:
Bike-shedding: move this out of the try/except block, `signames` variable could be easily removed here and below (unless you get a bonus for line counts then by all means leave it in)